moabb/contexts/base.py

```diff
@@ -37,66 +45,85 @@ def __init__(self, datasets, pipelines):
 
         # check pipelines
         if not isinstance(pipelines, dict):
-            raise(ValueError("pipelines must be a dict or a Pipeline instance"))
+            raise(ValueError("pipelines must be a dict"))
 
         for name, pipeline in pipelines.items():
             if not(isinstance(pipeline, BaseEstimator)):
                 raise(ValueError("pipelines must only contains Pipelines instance"))
         self.pipelines = pipelines
-
-
-class WithinSubjectContext(BaseContext):
-    """Within Subject evaluation Context.
-
-    Evaluate performance of the pipeline on each subject independently.
-
-    Parameters
-    ----------
-    datasets : List of Dataset instances.
-        List of dataset instances on which the pipelines will be evaluated.
-    pipelines : Dict of pipelines instances.
-        Dictionary of pipelines. Keys identifies pipeline names, and values
-        are scikit-learn pipelines instances.
-
-    See Also
-    --------
-    BaseContext
-    """
-
-    def evaluate(self, verbose=False):
-        """Evaluate performances
-
-        Parameters
-        ----------
-        verbose: bool (defaul False)
-            if true, print results durint the evaluation
-
-        Returns
-        -------
-        results: Dict of panda DataFrame
-            Return a dict of pandas dataframe, one for each pipeline
+        self.results = Results(type(evaluator).__name__, pipelines)
+
+    def verify(self, dataset):
+        assert(dataset.paradigm == 'imagery')
+
+    def process(self):
+        # Verify that datasets are valid for given paradigm first
+        for d in self.datasets:
+            self.verify(d)
+        for d in self.datasets:
+            print('\n\nProcessing dataset: {}'.format(d.code))
+            self.evaluator.preprocess_data(d, self)
+            for s in d.subject_list:
+                for name, clf in self.pipelines.items():
+                    self.results.add(self.process_subject(d, s, clf), name)
+        self.results.to_dataframe()
+
+    def process_subject(self, dataset, subj, clf):
+        return self.evaluator.evaluate(dataset, subj, clf, self)
+
+    def _epochs(self, raws, event_dict, time, channels=None):
+        '''Take list of raws and returns a list of epoch objects. Implements 
+        imagery-specific processing as well
+
+        '''
+        bp_low = self.fmin
```

Comment:
can you add fmin and fmax as parameters in `__init__`
if they are required in a function of the base class, they must be defined.